### ISSUE TITLE
Multiple errors in Postmodeling ...

### ISSUE THREAD

**Issue by nanounanue** [dssg/triage#638]
Using the following config file:

```yaml
# Postmodeling Configuration File

  project_path: '/triage' # Project path defined in triage with matrices and models
  audition_output_path: '/triage/audition/results_model_group_ids.json'



  thresholds: # Thresholds for2 defining positive predictions
        rank_abs: [50, 100, 250]
        rank_pct: [5, 10, 25]

  baseline_query: | # SQL query for defining a baseline for comparison in plots. It needs a metric and parameter
      select g.model_group_id,
             m.model_id,
             extract('year' from m.evaluation_end_time) as as_of_date_year,
             m.metric,
             m.parameter,
             m.value,
             m.num_labeled_examples,
             m.num_labeled_above_threshold,
             m.num_positive_labels
       from test_results.evaluations m
       left join model_metadata.models g
       using(model_id)
       where g.model_group_id = 65
             and metric = 'precision@'
             and parameter = '10_pct'

  max_depth_error_tree: 5 # For error trees, how depth the decision trees should go?
  n_features_plots: 10 # Number of features for importances
  figsize: [12, 12] # Default size for plots
  fontsize: 20 # Default fontsize for plots
```

I create the object `params`

```python
> params = PostmodelParameters('eis_postmodeling_config.yaml')
> params.__dict__

```

It returns this:

```python
{'project_path': '/triage',
 'audition_output_path': '/triage/audition/results_model_group_ids.json',
 'thresholds': {'rank_abs': [50, 100, 250], 'rank_pct': [5, 10, 25]},
 'baseline_query': "select g.model_group_id,\n       m.model_id,\n       extract('year' from m.evaluation_end_time) as as_of_date_year,\n       m.metric,\n       m.parameter,\n       m.value,\n       m.num_labeled_examples,\n       m.num_labeled_above_threshold,\n       m.num_positive_labels\n from test_results.evaluations m\n left join model_metadata.models g\n using(model_id)\n where g.model_group_id = 65\n       and metric = 'precision@'\n       and parameter = '10_pct'\n",
 'max_depth_error_tree': 5,
 'n_features_plots': 10,
 'figsize': (12, 12),
 'fontsize': 20,
 'model_group_id': [52, 48, 64, 52, 48, 64, 49, 53, 57, 48, 52, 56]}
```

**QUESTION** Why the last line has repeated model groups?

Following the postmodeling's example notebook:

```python
engine = create_pgconn('database.yaml')
list_tuple_models = get_models_ids(params.model_group_id, engine)
l_t = [ModelEvaluator(i.model_group_id, i.model_id, engine) for i in list_tuple_models]

# Model group object (useful to compare across model_groups and models in time)
audited_models_class = ModelGroupEvaluator(tuple(params.model_group_id), engine)
```

And then, everything falls apart:

```python
> audited_models_class
```

Throws the following error:

```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/usr/local/lib/python3.6/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

/usr/local/lib/python3.6/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    400                         if cls is not object \
    401                                 and callable(cls.__dict__.get('__repr__')):
--> 402                             return _repr_pprint(obj, self, cycle)
    403 
    404             return _default_pprint(obj, self, cycle)

/usr/local/lib/python3.6/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    695     """A pprint that just redirects to the normal repr function."""
    696     # Find newlines and replace them with p.break_()
--> 697     output = repr(obj)
    698     for idx,output_line in enumerate(output.splitlines()):
    699         if idx:

/usr/local/lib/python3.6/site-packages/triage/component/postmodeling/contrast/model_group_evaluator.py in __repr__(self)
    125     def __repr__(self):
    126         return (
--> 127            str({ 'Model_ids':{self.model_id},
    128                  'Model Groups': {self.model_group_id},
    129                  'Model types': {self.model_type},

TypeError: unhashable type: 'list'
```

This one is ok:

```python
> list_tuple_models
```

```python
[ModelEvaluator(model_group_id=48, model_id=138),
 ModelEvaluator(model_group_id=49, model_id=139),
 ModelEvaluator(model_group_id=52, model_id=142),
 ModelEvaluator(model_group_id=53, model_id=143),
 ModelEvaluator(model_group_id=56, model_id=146),
 ModelEvaluator(model_group_id=57, model_id=147),
 ModelEvaluator(model_group_id=64, model_id=154),
 ModelEvaluator(model_group_id=48, model_id=174),
 ModelEvaluator(model_group_id=49, model_id=175),
 ModelEvaluator(model_group_id=52, model_id=178),
 ModelEvaluator(model_group_id=53, model_id=179),
 ModelEvaluator(model_group_id=56, model_id=182),
 ModelEvaluator(model_group_id=57, model_id=183),
 ModelEvaluator(model_group_id=64, model_id=190),
 ModelEvaluator(model_group_id=48, model_id=210),
 ModelEvaluator(model_group_id=49, model_id=211),
 ModelEvaluator(model_group_id=52, model_id=214),
 ModelEvaluator(model_group_id=53, model_id=215),
 ModelEvaluator(model_group_id=56, model_id=218),
 ModelEvaluator(model_group_id=57, model_id=219),
 ModelEvaluator(model_group_id=64, model_id=226)]
```

But the object `l_t`:

```python
> l_t

l_t

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/usr/local/lib/python3.6/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

/usr/local/lib/python3.6/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    383                 if cls in self.type_pprinters:
    384                     # printer registered in self.type_pprinters
--> 385                     return self.type_pprinters[cls](obj, self, cycle)
    386                 else:
    387                     # deferred printer

/usr/local/lib/python3.6/site-packages/IPython/lib/pretty.py in inner(obj, p, cycle)
    561                 p.text(',')
    562                 p.breakable()
--> 563             p.pretty(x)
    564         if len(obj) == 1 and type(obj) is tuple:
    565             # Special case for 1-item tuples.

/usr/local/lib/python3.6/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    400                         if cls is not object \
    401                                 and callable(cls.__dict__.get('__repr__')):
--> 402                             return _repr_pprint(obj, self, cycle)
    403 
    404             return _default_pprint(obj, self, cycle)

/usr/local/lib/python3.6/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    695     """A pprint that just redirects to the normal repr function."""
    696     # Find newlines and replace them with p.break_()
--> 697     output = repr(obj)
    698     for idx,output_line in enumerate(output.splitlines()):
    699         if idx:

/usr/local/lib/python3.6/site-packages/triage/component/postmodeling/contrast/model_evaluator.py in __repr__(self)
    111              'model_group': self.model_group_id,
    112              'model_type': self.model_type,
--> 113              'as_of_date': self.as_of_date.strftime("%B %d, %Y"),
    114              'model_hyperparameters': self.hyperparameters})
    115         )

AttributeError: 'NoneType' object has no attribute 'strftime'

```

**Comment by nanounanue** [comment]
Pinging @ivanhigueram

**Comment by thcrock** [comment]
My thoughts:

1. Looking at the Model Group Evaluator error, I don't entirely understand why the arguments (that might be lists apparently) are turned into sets via the set construction literal for the __repr__. What if they were just raw?

2. What if we didn't do so much work in the __repr__? The __repr__ in both cases is actually depending on data that comes from self.metadata which originally comes from a query. What if we just made the repr reflect *user input* to the classes?

3. The ModelEvaluator error, at least, seems to originally come from the metadata query not returning any as-of-date. Provided that the model group ids and model ids are good, I'd say the most likely cause of this is the lack of predictions. Did you generate predictions for these models? As the code stands now, the predictions is most likely the root cause. However, based on that data that the metadata query is retrieving from predictions, there should be no reason why it *has* to get them from predictions. The commit to add matrix_uuid to the evaluations tables was added a month ago, so we should be able to rewrite the predictions part of that metadata query to bypass predictions entirely and grab that data from evaluations.

Again, the __repr__ is a nasty way to find out about this. The metadata cachedproperty should eagerly produce an error if it doesn't have all the data it needs.
